### PR TITLE
Develop: Prevent ClassNotFound exception for document triggers

### PR DESCRIPTION
**Pull request by dizzzz**
### Description:

Make sure 'correct' classloader is used; in current v5 situation (document) Trigger classes that are in a JAR file in a XAR file can not be found. Using the classloader of the BrokerPool fixes this issue

```
2019-08-10 23:32:47,522 [qtp728236551-42] WARN  (CollectionConfiguration.java [configureTrigger]:336) - Trigger class not found: org.exist.jms.replication.publish.ReplicationTrigger 
java.lang.ClassNotFoundException: org.exist.jms.replication.publish.ReplicationTrigger
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[?:1.8.0_212]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_212]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_212]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_212]
	at java.lang.Class.forName(Class.java:264) ~[?:1.8.0_212]
	at org.exist.collections.CollectionConfiguration.configureTrigger(CollectionConfiguration.java:309) ~[exist-core-5.0.0-RC9-SNAPSHOT.jar:5.0.0-RC9-SNAPSHOT]
```

### Reference:

[exist.log](https://slack-files.com/TG47XS7E2-FM8MAGFNU-29d292b843) 
[expath.log](https://slack-files.com/TG47XS7E2-FM87P4HE0-cf9226fb7d)
[xar](https://slack-files.com/TG47XS7E2-FLX8D6N10-6360eacc9c)

### Type of tests:

how to reproduce: 
- install messaging/replication XAR (snapshot)
- open eXide
- open /db/apps/messaging-replication/demo/replication/collection.xconf
- edit document, save + allow re-index

Outdated:
- no automated tests yet
- after fix an other issue appears where the parameters of the trigger are not shared with the trigger code. not sure if this is related to THIS specific class loader change

### CODE REVIEW

File: exist-core/src/main/java/org/exist/collections/CollectionConfiguration.java [hunk 1]
```diff
@@ -299,14 +299,14 @@ public IndexSpec getIndexConfiguration() {
         return indexSpec;
     }
 
-    private void configureTrigger(final Element triggerElement, final XmldbURI collectionConfigurationURI, final boolean testOnly) throws CollectionConfigurationException {
+    private void configureTrigger(ClassLoader cl, final Element triggerElement, final XmldbURI collectionConfigurationURI, final boolean testOnly) throws CollectionConfigurationException {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/collections/CollectionConfiguration.java]: Argument could be `final`.

Comment (dizzzz) [reply exist-core/src/main/java/org/exist/collections/CollectionConfiguration.java]: I Will work on both in 2-3 weeks....

**Comment by dizzzz** [comment]
I agree that a test is needed........ not sure yet how to do this...

**Comment by adamretter** [comment]
Hey @dizzzz if you need some help with the tests do you want to arrange a quick call with me?

**Comment by dizzzz** [comment]
The code of the 'example trigger' is in https://github.com/dizzzz/existdb-trigger-test/

**Comment by duncdrum** [comment]
Blegh Contents look good to me. Ci failures are a bit odd

File: exist-core/src/main/java/org/exist/collections/CollectionConfiguration.java (hunk 1 above)
Comment (dizzzz) [reply exist-core/src/main/java/org/exist/collections/CollectionConfiguration.java]: done

**Comment by dizzzz** [comment]
Ok, I found a FUN-damental class loader issue....

**Comment by joewiz** [comment]
TravisCI passed except macOS, which appears to be the same issue @duncdrum and @adamretter described on Monday's community call. Here's the error from the Travis macOS build:

> [ERROR] Failed to execute goal org.owasp:dependency-check-maven:5.2.1:check (default) on project exist-service: One or more exceptions occurred during dependency-check analysis: One or more exceptions occurred during analysis:
> [ERROR] 	Failed to request component-reports 

So I think we can say that the Travis macOS failure has nothing to do with this PR. 

Now, just waiting for AppVeyor...

**Comment by dizzzz** [comment]
All tests green!!!!

**Comment by joewiz** [comment]
@adamretter Now that @dizzzz has added a test and it is passing, are your concerns in https://github.com/eXist-db/exist/pull/2949#pullrequestreview-274790478 addressed?

**Comment by joewiz** [comment]
On today's Community Call @dizzzz notes that the "parameters not loaded" issue is resolved in the PR. So all known issues have been addressed here.

**Comment by wolfgangmm** [comment]
Merged as discussed in today's call. The PR is important and needed to further test replication with 5.0.

**Comment by adamretter** [comment]
@dizzzz Apologies, it seems that I did not press "submit" on the review that I previously conducted. At least the `scope` thing needs to be fixed ASAP otherwise the release build will fail on the dependency check.

**Pull request merged**